OCPBUGS-3356: E2E test for cookie length truncation

rfredette opened the pull request:
This PR adds a new test, `TestCookieLen`, which verifies that cookies in `spec.logging.access.httpCaptureCookies` are truncated to the specified max length.

This tests the fix in openshift/router#436

rfredette commented:
With this and the related router PR loaded on a cluster, I don't see the truncation warning, but I am seeing this message popping up every few seconds in the router logs: 
```
E1221 23:24:41.426700       1 webhook.go:90] Failed to make webhook authenticator request: the server could not find the requested resource
```
If I revert the ingress operator to the version without this change, I see the truncation warning but not this webhook message, so I don't think it's a configuration issue, but I need to dig some more to figure out what this error message even really means...

rfredette commented:
/jira refresh

candita commented on pkg/operator/controller/ingress/deployment.go:
```diff
@@ -820,12 +821,29 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
+			// is captured for every request and every response, and the
+			// maximum length is always allocated, so the memory impact is
+			// high.
+			if maxLength > 63 {
```
`maxLength` should be evaluated prior to setting `RouterCaptureHTTPCookies`, since it uses the `maxLength`.

rfredette commented:
/remove-lifecycle rotten

rfredette commented:
some tests failed due to an issue building the image, but since other tests didn't hit that issue, I'm assuming it was a timing thing.
/retest

rfredette commented:
The new test, `TestCookieLen`, won't pass in CI until openshift/router#436 merges. I ran the test on a cluster with both this PR and openshift/router#436, and it passes:
```$ TEST="TestCookieLen" make test-e2e
CGO_ENABLED=1 GO111MODULE=on GOFLAGS=-mod=vendor go test -timeout 1h -count 1 -v -tags e2e -run "TestCookieLen" ./test/e2e
=== RUN   TestCookieLen                                                                                                                                                       
=== PAUSE TestCookieLen                                                                
=== CONT  TestCookieLen                                                                                                                                                       
    tuning_options_test.go:516: Deleted cookielen-curl-6qf4c                   
    tuning_options_test.go:516: deleted ingresscontroller test-cookielen-2kt9d       
--- PASS: TestCookieLen (85.55s)                                                                                                                                              
PASS                                                                                   
ok      github.com/openshift/cluster-ingress-operator/test/e2e  86.000s      
```

rfredette commented:
After looking at this again, it's simpler to have the router figure out when to set `tune.http.cookielen` from the existing environment variables, rather than having the operator add a new environment variable. I've updated openshift/router#436 to do that, and now this PR only includes a new e2e test. Like before, the e2e test won't pass in CI until openshift/router#436 merges, but I've run it against a cluster with openshift/router#436 deployed, and it passes:
```
$ TEST="TestCookieLen" make test-e2e
CGO_ENABLED=1 GO111MODULE=on GOFLAGS=-mod=vendor go test -timeout 1h -count 1 -v -tags e2e -run "TestCookieLen" ./test/e2e
=== RUN   TestCookieLen
=== PAUSE TestCookieLen
=== CONT  TestCookieLen
    tuning_options_test.go:523: Deleted cookielen-curl-vcmdx
    tuning_options_test.go:523: deleted ingresscontroller test-cookielen-2llcg
--- PASS: TestCookieLen (84.51s)
PASS
ok      github.com/openshift/cluster-ingress-operator/test/e2e  84.808s
```

candita commented:
/assign @frobware

frobware commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,170 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+	defer assertIngressControllerDeleted(t, kclient, ic)
+
+	if err := waitForIngressControllerCondition(t, kclient, 2*time.Minute, icName, availableConditionsForPrivateIngressController...); err != nil {
+		t.Errorf("Timed out waiting for ingresscontroller %s to become available: %v", icName.Name, err)
```
If the ingress controller doesn't become ready then this should be a call to Fatal.

rfredette replied:
done

frobware commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,170 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+	}
+	defer assertDeleted(t, kclient, curlPod)
+
+	// Wait for the curl pod to be available
```
nit: seems like we wait for it to be Ready.

rfredette replied:
done

frobware commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,170 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+		t.Fatalf("Failed to get router deployment for ingresscontroller %s: %v", icName.Name, err)
+	}
+
+	curlPodImage := routerDeployment.Spec.Template.Spec.Containers[0].Image
```
We should first assert on the length of Spec.Containers == 1.

rfredette replied:
Rather than assert on `Spec.Containers = 1` (since if logging is set to "container" on the default ingresscontroller, there would be 2 containers), I updated it to loop through the containers and get the image from the one named "router" and `t.Fatalf` if that fails. Does that satisfy your concern?

frobware commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,170 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+				foundCookie = true
+				break
+			}
+			t.Errorf("Found cookie but length did not match. Expected %d, got %d", maxLength, len(cookieValue))
```
Should we continue to iterate over the remaining lines if this error hold true? Is this actually Fatal?

rfredette replied:
done

frobware reviewed (changes requested):
I'd like to see tests for where the maxLength for the cookie capture is 63 and 1024. The API forces a value so the first would exercise the default, and the latter tests for the API's maximum. In both cases we should assert that cookie values that are 1 more than the specified max get truncated.

rfredette commented on test/e2e/client_tls_test.go:
```diff
@@ -1412,19 +1412,6 @@ func verifyCRLs(t *testing.T, pod *corev1.Pod, expectedCRLs map[string]*x509.Rev
 	return true, nil
 }
 
-func getPods(t *testing.T, cl client.Client, deployment *appsv1.Deployment) (*corev1.PodList, error) {
```
done

rfredette commented:
I updated the test to add a few test cases, one at 63 (the previous limit), one at 64 (one over the limit to check the lower bound), one at 128, and one at 1024 (the max the API allows). The 1024 byte test is currently failing, but it seems like syslog is truncating the log line, because if I shorten the log format that haproxy is using, the cookie gets truncated at a larger value.

That said, I'm not sure I can change the log format enough to get the full 1024 bytes. I'm going to see if I can tune syslog to allow longer log lines, but if not it may not be a very useful test case.

frobware commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,170 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+			}
+		}
+		return false, nil
+	})
```
Looking at https://github.com/openshift/cluster-ingress-operator/pull/871/commits/6d68259cd2fb1996c89575493ffee66be773167f I didn't see an update  where we handle this error.

rfredette commented:
It looks like it wasn't syslog that was truncating the logs, but instead it was haproxy. When using the "Container" logging strategy, we currently leave `ROUTER_LOG_MAX_LENGTH` unset, and the template has a default log length of 1024 characters. By setting `ROUTER_LOG_MAX_LENGTH` to 2048 when using Container logging, the new test is passing on my cluster, so I've included that commit here.

I imagine, though, that the better way to solve that is to update the template in openshift/router instead, but some discussion is warranted on what to update the default to.

rfredette commented:
must gather failure
/retest

rfredette commented:
aws-operator failed on cluster setup
/retest

rfredette commented:
Another must-gather failure
/retest

frobware commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,195 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+				}
+			}()
+
+			// Search the logs container for the specified cookie, making sure that it exactly matches the expected truncated value.
+			expectedCookieString := fmt.Sprintf("%s=%s", cookieName, padding)
+			cookieNameRegexp := regexp.MustCompile(fmt.Sprintf("cookie: (%s=a+x?)$", cookieName))
+			scanner := bufio.NewScanner(readCloser)
+			for scanner.Scan() {
+				line := scanner.Text()
+				// FindSubmatch returns a slice of []byte. Index 0 contains the entire matched string, and each following index
+				// contains one of the submatches (sections of the regexp in parentheses). There's only 1 submatch in this
+				// regexp, so it should either return nil if there's no match, or a 2-item slice.
+				if match := cookieNameRegexp.FindSubmatch([]byte(line)); match != nil {
+					if len(match) != 2 {
+						// It shouldn't be possible to match the regexp without the submatch, but this check will cover our
+						// bases.
+						t.Fatalf("Found regexp match but not submatch")
+					}
+					cookieValue := match[1]
+					if len(cookieValue) == maxLength {
+						return
+					}
+					t.Fatalf("Found cookie but length did not match. Expected %d, got %d", maxLength, len(cookieValue))
+				}
+			}
+			t.Errorf("Cookie %q was not found in router pod %q logs", expectedCookieString, routerPod.Name)
```
```suggestion
			// Search the logs container for the specified
			// cookie, making sure that it exactly matches
			// the expected truncated value.
			expectedTruncatedCookieString := fmt.Sprintf("%s=%s", cookieName, padding) // this is without the 'x'
			foundExpectedCookie := false
			scanner := bufio.NewScanner(readCloser)
			for scanner.Scan() {
				line := strings.TrimSpace(scanner.Text())
				if strings.Contains(line, cookieString) {
					t.Fatalf("found untruncated cookie string: %v", cookieString)
				}
				if strings.Contains(line, expectedTruncatedCookieString) {
					foundExpectedCookie = true
				}
			}
			if !foundExpectedCookie {
				t.Fatalf("did not find expected truncated cookie string: %v", expectedTruncatedCookieString)
			}
```

frobware replied:
I changed the test to also assert that we don't match the cookie that has the trailing "x" in addition to asserting that there's a match for the truncated value.

frobware replied:
If it's easier to consume then here's my suggestion with the addition that we check the scanner for errors too:

```go
			// Search the logs container for the specified
			// cookie, making sure that it exactly matches
			// the expected truncated value.
			expectedTruncatedCookieString := fmt.Sprintf("%s=%s", cookieName, padding) // this is without the 'x'
			foundExpectedCookie := false
			scanner := bufio.NewScanner(readCloser)
			for scanner.Scan() {
				line := strings.TrimSpace(scanner.Text())
				if strings.Contains(line, cookieString) {
					t.Fatalf("found untruncated cookie string: %v", cookieString)
				}
				if strings.Contains(line, expectedTruncatedCookieString) {
					foundExpectedCookie = true
				}
			}
			if err := scanner.Err(); err != nil {
				t.Fatalf("Error while reading container logs: %v", err)
			}
			if !foundExpectedCookie {
				t.Fatalf("did not find expected truncated cookie string: %v", expectedTruncatedCookieString)
			}
```

rfredette replied:
Thanks, this does make it much more straightforward. I've incorporated it in my latest commit.

frobware commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,195 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+		name      string
+		maxLength int
+	}{
+		{"63 bytes", 63},
+		{"64 bytes", 64},
+		{"128 bytes", 128},
+		{"1024 bytes", 1024},
```
We need an explanation for each of the magic numbers. 

```go
	testCases := []struct {
		name      string
		maxLength int
	}{
		// 63 is one less than the default maximum length of a
		// cookie (64). We test with this length to ensure our
		// system handles values just below the default
		// maximum correctly.
		{"63 bytes", 63},

		// 64 is the default maximum length for a cookie
		// value. We test with this length to ensure our
		// system behaves correctly when the default maximum
		// length is used.
		{"64 bytes", 64},

		// 128 is significantly more than the default maximum
		// length. We test with this length to see how our
		// system behaves when a value larger than the default
		// maximum is used.
		{"128 bytes", 128},

		// 1024 is the maximum length of a cookie as defined
		// by the OpenShift Ingress API. We test with this
		// length to ensure our system can handle the largest
		// cookie value that OpenShift allows.
		{"1024 bytes", 1024},
	}
```

rfredette replied:
Do you care if it's in the test case name, or will a comment suffice?

frobware commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,195 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			maxLength := testCase.maxLength
+			t.Parallel()
```
Out of consistency please can we move this to the first line in the subtest (i.e., before `maxLength`).

candita commented:
/assign

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,195 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+				Name:      names.SimpleNameGenerator.GenerateName("cookielen-curl-"),
+				Namespace: "default",
+			}
+			curlPod := buildExecPod(curlPodName.Name, curlPodName.Namespace, curlPodImage)
```
Why not use `buildCurlPod`?

rfredette replied:
Good question. `buildCurlPod` does look like it would simplify things; I'll rework it to use that.

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,195 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+				Name:      names.SimpleNameGenerator.GenerateName("test-cookielen-"),
+			}
+			domain := icName.Name + "." + dnsConfig.Spec.BaseDomain
+			ic := newPrivateController(icName, domain)
```
Is there a reason you need to create a new ic for each test case?

rfredette replied:
The cases run in parallel, so unless I'm mistaken, they each need their own ingress controller.

candita replied:
I was thinking - create one ic, and create one pod per test case.  That shouldn't cause any conflicts that I can think of.  I'm a little concerned that there could be resource failures or timeouts creating so many controllers.

rfredette replied:
I don't think one ic would work; Each test case sets the ic's `spec.logging.access.httpCaptureCookies.maxLength` field to a different value, and you can only set one cookie to capture per ic.

candita replied:
Normally we run each test in parallel, but here you are running each test case in parallel.  Did you get advice to do it this way?  I don't see the advantages, but do see the aforementioned potential disadvantage in that resource use could be compromised by running 4 private ics in parallel in ci.

Could you at least add a comment and note that it might cause an issue and point out how to fix that issue (don't run in parallel and use one private ic).

frobware replied:
We create many ingresscontoller's in the e2e tests and creating 4 more shouldn't be a problem (until some limit proves me wrong). 

If we created 4 new discrete TestFeature functions, each creating their own private IC, to accommodate and test new features over subsequent OCP releases would you have the same concerns? 

w.r.t running each test case in parallel--that seems fine to me. It's really no different to creating 4 new TestX functions that use `t.Parallel`.

rfredette commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,195 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+			routerPodList, err := getPods(t, kclient, routerDeployment)
+			if err != nil {
+				t.Fatalf("failed to get pods in deployment %s: %v", routerDeploymentName, err)
+			} else if len(routerPodList.Items) == 0 {
```
done

rfredette commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +327,195 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+
+			// Search the logs container for the specified cookie, making sure that it exactly matches the expected truncated value.
+			expectedCookieString := fmt.Sprintf("%s=%s", cookieName, padding)
+			cookieNameRegexp := regexp.MustCompile(fmt.Sprintf("cookie: (%s=a+x?)$", cookieName))
```
This should be addressed with the rework in the latest commit.

rfredette commented:
After testing the latest changes, I ran into a few issues, one with the conversion to use `buildCurlPod`, and another where the canary's `spec.host` field was empty, which caused the test to try to curl `""`. Both have been fixed in the commits I just pushed.

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +326,201 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+			// Send a request through the router that includes the captured cookie, and verify that it's truncated as expected.
+
+			// Use the router image for the client pod since it includes curl.
+			routerDeployment := &appsv1.Deployment{}
```
Could you take a quick look at how `buildCurlPod` is done in https://github.com/openshift/cluster-ingress-operator/blob/master/test/e2e/operator_test.go#L3093 `TestUniqueIDHeaders`?  I think you should build your own simple image, route, and service rather than use the ones provided by router, canary and ingress.

rfredette replied:
The advantage to using the canary route is that it's guaranteed to already be deployed, which speeds up the test by not requiring additional pods and routes to be created. That said, it does make the test more confusing for a pretty small gain. I'll rework this to use an echo pod like `TestUniqueIDHeaders`.

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +326,201 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+		// 63 bytes was the limit before OCPBUGS-3356 was fixed; This should pass even without the patch, so it's more
+		// of a sanity check.
+		{"63 bytes", 63},
+		// The lower bound. 64 bytes is the smallest number that would be truncated short prior to the OCPBUGS-3356 fix.
```
How can both 63 and 64 be the lower limit?

rfredette replied:
63 bytes was the limit before the fix in openshift/router#436. You could remove that case and it would cover the range where behavior has changed due to the router PR, but I think it's helpful to include the 63 byte case as a sanity check. 
If it passes but all the other cases fail, there's something wrong with the way the global variable `tune.http.cookielen` was set, but if it and the other cases fail, it indicates that the failure is probably unrelated to `tune.http.cookielen`.

candita replied:
I'm going to suggest some re-wording of the comments now that I've reviewed https://github.com/openshift/router/pull/436

rfredette replied:
Thanks for your suggestions. I did adjust the wording slightly, but I removed any reference to the specific jira issue. Let me know if you think they are in need of any additional edits.

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +326,201 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+			curlArgs := []string{
+				"-k", "-L",
+				"-b", cookieString,
+				// --resolve used this way makes curl send the request to our ingress controller's internal service address as
+				// if it was directed to it by a load balancer. This ensures the request is handled by our ingress controller
+				// without having to wait extra time for a load balancer to be provisioned.
+				"--resolve", fmt.Sprintf("%s:443:%s", canaryRouteURL, icInternalService.Spec.ClusterIP),
```
I don't think you need to do this at all. Can you explain?  It's better to use the established pattern mentioned in https://github.com/openshift/cluster-ingress-operator/pull/871/files#r1296550078, unless there is a need to do it differently.

rfredette replied:
`TestUniqueIdHeader` does the same thing here: https://github.com/openshift/cluster-ingress-operator/blob/master/test/e2e/operator_test.go#L3163 and there are a few other test cases that use this same behavior. It bypasses DNS and explicitly sends the request to the router pod(s) for a particular ingress controller, guaranteeing that the right router pod handles the connection, and removing any chance of slow DNS propagation causing test delays or failures.

candita commented on test/e2e/util_test.go:
```diff
@@ -137,7 +137,11 @@ func buildCurlPod(name, namespace, image, host, address string, extraArgs ...str
 	}
 	curlArgs = append(curlArgs, extraArgs...)
-	curlArgs = append(curlArgs, "http://"+host)
+	if strings.HasPrefix(host, "http://") || strings.HasPrefix(host, "https://") {
```
I think the only reason you need to do this is because you're borrowing the canary route.  Please don't borrow the canary route, as mentioned previously.

rfredette replied:
I think it would be good for `buildCurlPod` to allow https, but with the switch to an echo pod, it doesn't need to be done here, so I've removed this change.

rfredette commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +326,201 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+		name      string
+		maxLength int
+	}{
+		// 63 bytes was the limit before OCPBUGS-3356 was fixed; This should pass even without the patch, so it's more
```
"The patch" here means the fix for OCPBUGS-3356, but I'll reword it for better clarity.

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +326,201 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+					foundExpectedCookie = true
+				}
+			}
+			if err := scanner.Err(); err != nil {
+				t.Fatalf("Error while reading container logs: %v", err)
+			}
```
If you found the right cookie, you probably wouldn't want this to be a Fatal error.  
```suggestion
			if err := scanner.Err(); err != nil {
				t.Errorf("Error while reading container logs: %v", err)
			}
```

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +326,201 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+			}
+			if !foundExpectedCookie {
+				t.Fatalf("did not find expected truncated cookie string: %v", expectedTruncatedCookieString)
+			}
```
Add a message here so that if there are some failures and some successes, you will know which succeeded.
```suggestion
			}
			t.Logf("found expected cookie truncated to %d", maxLength)
```

rfredette replied:
Each test case is run with `t.Run()`, so they show up separately in the test output. To induce a failure and to test out a question you asked in another comment, I changed the 1024 byte case to 1026 bytes and ran the test. Here's a snippet of the output:
```
=== NAME  TestCookieLen/1026_bytes
    tuning_options_test.go:379: Failed to create ingresscontroller openshift-ingress-operator/test-cookielen-hv8vz: IngressController.operator.openshift.io "test-cookielen-hv8vz" is invalid: spec.logging.access.httpCaptureCookies[0].maxLength: Invalid value: 1026: spec.logging.access.httpCaptureCookies[0].maxLength in body should be less than or equal to 1024

... cut for brevity ...

=== NAME  TestCookieLen/128_bytes
    tuning_options_test.go:541: deleted ingresscontroller test-cookielen-j5lxj
=== NAME  TestCookieLen/63_bytes
    tuning_options_test.go:541: deleted ingresscontroller test-cookielen-kh5rl
--- FAIL: TestCookieLen (0.00s)
    --- FAIL: TestCookieLen/1026_bytes (0.07s)
    --- PASS: TestCookieLen/64_bytes (90.57s)
    --- PASS: TestCookieLen/128_bytes (91.78s)
    --- PASS: TestCookieLen/63_bytes (93.00s)
FAIL
FAIL    github.com/openshift/cluster-ingress-operator/test/e2e  93.475s
FAIL
```
The output does get interleaved due to the parallel execution, but I think it's clear enough which test case is which.

candita replied:
I was asking about adding a positive message to match the negative message: "did not find expected truncated cookie string". It's a nit.

But the error message I see above for 1026 bytes, that seems weird.  It starts with "Failed to create ingresscontroller".  Can you add a test that you expect to fail and print out the reason why it fails rather than the reason why the ingress controller doesn't get created?

rfredette replied:
The error message in the 1026 case was the expected one. `httpCaptureCookies.maxLength` is a field of the ingresscontroller object, and the openshift API limits it to a number between 1 and 1024, so trying to set it to a number outside that range will cause the ingresscontroller to be rejected

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +326,201 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+		name      string
+		maxLength int
+	}{
+		// 63 bytes was the limit before OCPBUGS-3356 was fixed; This should pass even without the patch, so it's more
+		// of a sanity check.
```
```suggestion
		// 63 bytes is the HAProxy default, which would be the case when no maxLength is specified
```

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +326,201 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+		// 63 bytes was the limit before OCPBUGS-3356 was fixed; This should pass even without the patch, so it's more
+		// of a sanity check.
+		{"63 bytes", 63},
+		// The lower bound. 64 bytes is the smallest number that would be truncated short prior to the OCPBUGS-3356 fix.
```
```suggestion
		// The lower bound. 64 bytes is the smallest number above the HAProxy default, and for which a HAProxy config change is made.
```

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +326,201 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+		{"63 bytes", 63},
+		// The lower bound. 64 bytes is the smallest number that would be truncated short prior to the OCPBUGS-3356 fix.
+		{"64 bytes", 64},
+		// 128 bytes is just a number between the upper and lower bounds. If this passes and either boundary cases fail,
+		// it's likely an off-by-one or similar boundary issue.
```
```suggestion
		// 128 bytes is an additional test to prove that the truncation length is correct.
```

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +326,201 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+		// 128 bytes is just a number between the upper and lower bounds. If this passes and either boundary cases fail,
+		// it's likely an off-by-one or similar boundary issue.
+		{"128 bytes", 128},
+		// The upper bound. 1024 is the maximum allowed by the API.
```
What would happen for 1026?  Would it default to 1024 or 63, or break the test?

rfredette commented:
Test failures are unrelated.
/retest

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +325,219 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+				t.Fatalf("Failed to read logs from pod %s container %s: %v", routerPod.Name, logsContainerName, err)
+			}
+
+			defer func() {
+				if err := readCloser.Close(); err != nil {
+					t.Logf("Failed to close reader for pod %s container %s: %v", routerPod.Name, logsContainerName, err)
+				}
+			}()
```
I think this defer function needs to go after L525, where the readCloser is used.

rfredette replied:
I don't think it matters, as long as the defer function is registered between `readCloser` being instantiated and the first possible return condition, since it won't get called until `TestCookieLen` returns.

candita commented on test/e2e/tuning_options_test.go:
```diff
@@ -320,3 +325,219 @@ func TestHAProxyTimeoutsRejection(t *testing.T) {
+			}
+
+			logsContainerName := "logs"
+			readCloser, err := client.CoreV1().Pods(routerPod.Namespace).GetLogs(routerPod.Name, &corev1.PodLogOptions{
```
The pattern that is used for scanning logs usually has a poll loop, like https://github.com/openshift/cluster-ingress-operator/blame/master/test/e2e/http_header_buffer_test.go#L134-L163.  If it isn't there the first time you check, it may not have arrived yet.

rfredette commented:
/retest

frobware commented:
/lgtm

candita commented:
/approve

candita commented:
/acknowledge-critical-fixes-only

candita commented:
/label acknowledge-critical-fixes-only
